### PR TITLE
[HW][Multithreading] Thread Pool

### PR DESCRIPTION
**Pull request by FedotovArtemSWTecNN**

### CODE REVIEW

File: ThreadPool/ThreadPool/ThreadPool/ThreadPool.h [hunk 1]
```diff
@@ -1,12 +1,29 @@
...
+#include <thread>
+#include <deque>
+#include <set>
+#include <atomic>
+#include <mutex>
+#include <condition_variable>
+
+// ThreadPool.cpp
+#include <algorithm>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (FedotovArtemSWTecNN) [reply ThreadPool/ThreadPool/ThreadPool/ThreadPool.h]: Headers moved

File: ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+static std::shared_ptr<ThreadPool> gs_threadPool{ nullptr };
+
+// Push a task to thread pool execution queue
+	// Returns false if task wasn't queued due to thread pool wasn't initialized
+	// Returns false if task wasn't queued due conflicting label
+	// Returns true, if task successfully queued
+bool ThreadPool::queue(const std::string &label, task_type task)
+{
+	auto lock = std::unique_lock<std::mutex>(c_queueMutex);
```
[the remark this answers is not in the document]
Comment (FedotovArtemSWTecNN) [reply ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp]: Fixed (I hope I understood the remark correctly)

File: ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+	for (std::size_t id = 0; id < poolSize; ++id)
+	{
+		c_threads.emplace_back([this, id]
+		{
+			while (true)
+			{
+				Task task;
+
+				{   // {} for auto unlock
+					auto lock = std::unique_lock<std::mutex>(c_queueMutex);
```
[the remark this answers is not in the document]
Comment (FedotovArtemSWTecNN) [reply ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp]: Fixed

File: ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+					}
+
+					task = c_tasks.front();
+					c_tasks.pop_front();
+					c_runningTasks.emplace(task.label);
+				}
+
+				task.callable();
+
+				auto lock = std::unique_lock<std::mutex>(c_queueMutex);
```
[the remark this answers is not in the document]
Comment (FedotovArtemSWTecNN) [reply ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp]: Fixed: lock() and unlock() methods used (I hope I understood the remark correctly)

File: ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+}
+
+
+// Wait till task with name <label> will be completed
+void ThreadPool::waitForTask(const std::string &label)
+{
+	std::cout << "Wait for task \"" + label + "\"" << std::endl;
+
+	std::unique_lock<std::mutex> lock(c_queueMutex);
+	c_queueConVar.wait(lock, [this, &label] { return !gs_threadPool || c_runningTasks.end() == c_runningTasks.find(label); });
```
Comment (an-ovsyuhno) [inline ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp:49]: I guess, we still need to check if our task is in c_tasks here (queued, but not started yet)?

File: ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+
+// Wait till task with name <label> will be completed
+void ThreadPool::waitForTask(const std::string &label)
+{
+	std::cout << "Wait for task \"" + label + "\"" << std::endl;
+
+	std::unique_lock<std::mutex> lock(c_queueMutex);
+	if (c_tasks.end() != std::find_if(c_tasks.begin(), c_tasks.end(), label))
```
Comment (an-ovsyuhno) [inline ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp:49]: I guess, this condition needs to be moved inside condition variable wait condition.

Comment (FedotovArtemSWTecNN) [reply ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp:49]: > I guess, this condition needs to be moved inside condition variable wait condition.

Yes, I'm sorry. I hurried to make commit. I'll fix it now

Comment (FedotovArtemSWTecNN) [reply ThreadPool/ThreadPool/ThreadPool/ThreadPool.cpp]: Fixed. I think it's ready. I didn't understand right away. I hope I got it right

**Pull request closed without merging**